### PR TITLE
Add limit to task payload size

### PR DESCRIPTION
**Pull request by georgew5656**
Currently it's possible to submit a really big task payload and have it OOM the overlord if the request fails and logs the whole thing in SQLMetadataStorageActionHandler.insertEntryWithHandle.

If the request happens to be larger than max_allowed_packet for a mysql metadata store it will always fail. Since really large task payloads seem to cause overlord instability in general I think it makes sense to limit the size of task payloads at the task queue level.

### Description
Add a new config that sets a limit for task payload sizes. Throw a exception if the limit is exceeded.

The default limit of 60 MB is based on the 64 MB default value of max_allowed_packet in MySQL 8+.

I would ideally like to use the http request content-length header to calculate the size of the task payload rather than re-serializing it in memory but we also call taskQueue.add directly from supervisors so that would bypass the check. If others think this is acceptable and it would be better to check Content-Length in OverlordResource I am fine with changing this logic.

#### Release note
- Adding a new guardrail for submitting tasks that are too large

##### Key changed/added classes in this PR
 * `TaskQueue`

This PR has:

- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ X added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by kfaraz** [comment]
@georgew5656 , thanks for the changes!

It makes sense to fail early on large payloads with a clear error message, but I have some concerns.

- Such tasks already fail on insertion into the metadata store. Do you think the current error message in this case is not clear enough? Instead of a new check/config, should we try to improve the existing error message by maybe catching it and throwing a cleaner `DruidException`?
- Or do you feel that there are other challenges with large task payloads too (say, Overlord facing a memory crunch)?
- The only problem with introducing this new config with a default of 64MiB is that some clusters would already be using large task payloads and would have set their `max_allowed_packet` to large enough values that allow such payloads. And if we go with a default that is large enough to accommodate all clusters, then we would probably defeat the purpose of the config itself.
- There is also the MSQ case to consider where the user just submits a SQL and the task payload is actually generated by the system. How does the user control the task payload size in that case? Maybe we don't fail in the case the of msq task types?

One compromise could be to keep the check and just _warn_ about large task payloads for the time being, with a log message and maybe also an alert event.

Let me know what you think!

**Comment by georgew5656** [comment]
> @georgew5656 , thanks for the changes!
> 
> It makes sense to fail early on large payloads with a clear error message, but I have some concerns.
> 
> * Such tasks already fail on insertion into the metadata store. Do you think the current error message in this case is not clear enough? Instead of a new check/config, should we try to improve the existing error message by maybe catching it and throwing a cleaner `DruidException`?
> * Or do you feel that there are other challenges with large task payloads too (say, Overlord facing a memory crunch)?
> * One problem with introducing this new config with a default of 64MiB is that some clusters would already be using large task payloads and would have set their `max_allowed_packet` to large enough values that allow such payloads. And if we go with a default that is large enough to accommodate all clusters, then we would probably defeat the purpose of the config itself.
> * There is also the MSQ case to consider where the user just submits a SQL and the task payload is actually generated by the system. How does the user control the task payload size in that case? Maybe we don't fail in the case of msq task types?
> 
> One compromise could be to keep the check and just _warn_ about large task payloads for the time being, with a log message and maybe also an alert event.
> 
> Let me know what you think!

the direct problem seems to be the error message that gets thrown when the write to db fails is a little TOO descriptive in this case (it tries to log the whole thing and this can easily cause mem issues).

i don't think the config totally loses its purpose with a higher default, since you can still choose to lower it to limit the blast radius of large payloads.

maybe it would make sense to just truncate that error message and throw a warn in OverlordResource when a task payload is very large? because I don't think it a task payload above a certain size makes sense

**Comment by georgew5656** [comment]
> > maybe it would make sense to just truncate that error message and throw a warn in OverlordResource when a task payload is very large?
> 
> Yes, we should definitely do these two items even if we decide not to do the rest.
> 
> > i don't think the config totally loses its purpose with a higher default, since you can still choose to lower it to limit the blast radius of large payloads.
> 
> True, but I don't think most cluster operators would think of updating this config when they encounter a problem. They would be more likely to increase the `max_allowed_packet` unless we update the error message thrown for `max_allowed_packet` to suggest using the new config.
> 
> > I personally don't think a task payload above a certain size makes sense. for msq, would it really try to generate that large of a payload? 60 MB is huge for metadata
> 
> I agree that 60MiB is large enough, but I do recall some cases where users had to resort to increasing the `max_allowed_packet`. Example,
> 
> ```
> Could not send query: query size is >= to max_allowed_packet (67108864)
> ```
> 
> In conclusion, we could do the following:
> 
> * Truncate the error message thrown when we exceed `max_allowed_packet` and include info about using the new config.
> * Log a warning when task payload crosses say 80% of max limit.
> * Log a stronger warning and raise an alert when task payload crosses 100% of max limit, maybe also mentioning that such task payloads will be rejected in future Druid releases.

i'm not sure if a config is needed in that case? I just added a hardcoded value in the warn check

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 1]
```diff
@@ -509,6 +509,22 @@ public boolean add(final Task task)
...
+    catch (JsonProcessingException ignored) {
+    }
+    if (payload != null && payload.length() > config.getMaxTaskPayloadSize()) {
+      log.warn("Received a task payload > [%d] with id [%s]. and datasource [%s]" +
+              " There may be downstream issues caused by managing this large payload." +
+              "Increase druid.indexer.queue.maxTaskPayloadSize to ignore this warning.",
+          config.getMaxTaskPayloadSize(),
+          task.getId(),
+          task.getDataSource()
+      );
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:525]:
```suggestion
      log.warn(
             "Task[%s] of datasource[%s] has payload size[%d] larger than the recommended maximum[%d]." +
              + " Large task payloads may cause stability issues in the Overlord and may fail while persisting to the metadata store."
              + " Use smaller task payloads or increase 'druid.indexer.queue.maxTaskPayloadSize' to suppress this warning.",
          task.getId(), task.getDataSource(), config.getMaxTaskPayloadSize()
      );
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/TaskQueueConfig.java [hunk 2]
```diff
@@ -42,20 +42,27 @@ public class TaskQueueConfig
...
+
   @JsonCreator
   public TaskQueueConfig(
       @JsonProperty("maxSize") final Integer maxSize,
       @JsonProperty("startDelay") final Period startDelay,
       @JsonProperty("restartDelay") final Period restartDelay,
       @JsonProperty("storageSyncRate") final Period storageSyncRate,
-      @JsonProperty("taskCompleteHandlerNumThreads") final Integer taskCompleteHandlerNumThreads
+      @JsonProperty("taskCompleteHandlerNumThreads") final Integer taskCompleteHandlerNumThreads,
+      @JsonProperty("maxTaskPayloadSize") final Long maxTaskPayloadSize
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/config/TaskQueueConfig.java:55]: We should also annotate this with `@Nullable`.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 3]
```diff
@@ -509,6 +509,22 @@ public boolean add(final Task task)
       throw EntryAlreadyExists.exception("Task[%s] already exists", task.getId());
     }
 
+    String payload = null;
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:512]: Move the whole new logic to a separate private method `validateTaskPayload()`.

File: docs/configuration/index.md [hunk 4]
```diff
@@ -1125,6 +1125,7 @@ These Overlord static configurations can be defined in the `overlord/runtime.pro
 |`druid.indexer.queue.startDelay`|Sleep this long before starting Overlord queue management. This can be useful to give a cluster time to re-orient itself (for example, after a widespread network issue).|`PT1M`|
 |`druid.indexer.queue.restartDelay`|Sleep this long when Overlord queue management throws an exception before trying again.|`PT30S`|
 |`druid.indexer.queue.storageSyncRate`|Sync Overlord state this often with an underlying task persistence mechanism.|`PT1M`|
+|`druid.indexer.queue.maxTaskPayloadSize`|Maximum size in bytes of a single task payload.|`62914560`|
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply docs/configuration/index.md:1128]: If we decide to specify the config in human readable bytes, then we can retain the name `maxTaskPayloadSize`.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 5]
```diff
@@ -1018,4 +1022,37 @@ List<Task> getTasks()
...
+                payload.length(),
+                config.getMaxTaskPayloadSize()
+            );
+      }
+    }
+    catch (JsonProcessingException e) {
+      throw DruidException.defensive(
+          "Failed to parse task payload for validation",
+          e
+      );
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1055]: `.defensive()` doesn't take a cause. You can just log the original exception here and not include any root cause in the final `DruidException` as it is a defensive exception anyway.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 6]
```diff
@@ -1018,4 +1022,37 @@ List<Task> getTasks()
       giant.unlock();
     }
   }
+
+  void validateTaskPayload(Task task)
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1026]: Can this be private?

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 7]
```diff
@@ -1018,4 +1022,37 @@ List<Task> getTasks()
...
+  {
+    try {
+      String payload = passwordRedactingMapper.writeValueAsString(task);
+      if (payload.length() > TASK_SIZE_WARNING_THRESHOLD) {
+        log.warn("Received a large task payload [%s] with id [%s] and datasource [%s]" +
+                " There may be downstream issues caused by managing this large payload." +
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1032]: Please update the messaging based on the older suggestion.

File: indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskQueueTest.java [hunk 8]
```diff
@@ -214,6 +216,73 @@ public void testAddThrowsExceptionWhenQueueIsFull()
     );
   }
 
+  @Test
+  public void testAddThrowsExceptionWhenPayloadIsTooLarge()
+  {
+    HumanReadableBytes maxPayloadSize = HumanReadableBytes.valueOf(10 * 1024 * 1024);
```
Comment (kfaraz) [inline indexing-service/src/test/java/org/apache/druid/indexing/overlord/TaskQueueTest.java:222]:
```suggestion
    HumanReadableBytes maxPayloadSize10Mib = new HumanReadableBytes("10MiB");
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 9]
```diff
@@ -1018,4 +1022,37 @@ List<Task> getTasks()
       giant.unlock();
     }
   }
+
+  void validateTaskPayload(Task task)
+  {
+    try {
+      String payload = passwordRedactingMapper.writeValueAsString(task);
+      if (payload.length() > TASK_SIZE_WARNING_THRESHOLD) {
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1030]: this if should be in the `else` of the second if. i.e.

```
if (maxAllowed != null && payloadSize > maxAllowed) {
    throw exception;
} else if (payloadSize > warningSize) {
    log warning/raise alert
}
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 10]
```diff
@@ -1018,4 +1022,37 @@ List<Task> getTasks()
...
+
+      if (config.getMaxTaskPayloadSize() != null && config.getMaxTaskPayloadSize().getBytesInInt() < payload.length()) {
+        throw DruidException.forPersona(DruidException.Persona.OPERATOR)
+            .ofCategory(DruidException.Category.INVALID_INPUT)
+            .build(
+                "Task payload size was [%d] but max size is [%d]. " +
+                    "Reduce the size of the task or increase 'druid.indexer.queue.maxTaskPayloadSize'.",
+                payload.length(),
+                config.getMaxTaskPayloadSize()
+            );
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1048]: Rephrased, used short-hand for invalid input exception, included task id in error message:

```suggestion
        throw InvalidInput.exception(
                "Task[%s] has payload of size[%d] but max allowed size is [%d]. " +
                    "Reduce the size of the task payload or increase 'druid.indexer.queue.maxTaskPayloadSize'.",
                task.getId(), payload.length(), config.getMaxTaskPayloadSize()
            );
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java (hunk 7 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1032]: Also, should we continue to warn if someone has explicitly set the `maxTaskPayloadSize` to be larger than the `TASK_SIZE_WARNING_THRESHOLD`?  I guess it makes sense to always give out a warning to encourage users/operators to reduce the task payload sizes.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 11]
```diff
@@ -1018,4 +1023,34 @@ List<Task> getTasks()
...
+      if (config.getMaxTaskPayloadSize() != null && config.getMaxTaskPayloadSize().getBytesInInt() < payload.length()) {
+        throw InvalidInput.exception(
+                "Task[%s] has payload of size[%d] but max allowed size is [%d]. " +
+                    "Reduce the size of the task payload or increase 'druid.indexer.queue.maxTaskPayloadSize'.",
+                task.getId(), payload.length(), config.getMaxTaskPayloadSize()
+            );
+      } else if (payload.length() > TASK_SIZE_WARNING_THRESHOLD) {
+        log.warn(
+            "Task[%s] of datasource[%s] has payload size[%d] larger than the recommended maximum[%d]. " +
+                "Large task payloads may cause stability issues in the Overlord and may fail while persisting to the metadata store." +
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1040]: Typo: missing space
```suggestion
                "Large task payloads may cause stability issues in the Overlord and may fail while persisting to the metadata store. " +
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java [hunk 12]
```diff
@@ -1018,4 +1023,34 @@ List<Task> getTasks()
...
+                "These large payloads will be rejected by the overlord in the future.",
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java:1041]:
```suggestion
                "Such tasks may be rejected by the Overlord in future Druid versions.",
```

**Review by kfaraz — APPROVED** [review]
Minor nitpicks, otherwise looks good.
Please update the release note in the PR description mentioning the usage of the new config.

**Comment by georgew5656** [comment]
the intellij inspection failures look unrelated to me

**Pull request merged**